### PR TITLE
Fix footer height calculation

### PR DESCRIPTION
**Pull request by lpeancovschi**
Fixes footer height calculation.

Without this fix footer height is equal to header height. That means that if you don't have a header but do have a footer your content will overlap the footer because header height is zero.

### CODE REVIEW

**Comment by philprime** [comment]
Hi, thanks for your contribution.

After a quick look I need you to provide an example to reproduce the issue, as your fix seems wrong:

The method you changed calculates the distance from the top edge, therefore it is paper_height - footer_height.

I could be wrong, that's why an example would be great

**Comment by lpeancovschi** [comment]
Hello @philprime,

Thanks for looking at this! The issue is reproducible when you have a table view that is rendered over a footer. To reproduce it you can paste this method to the [HeaderFooterExampleFactory.swift](https://github.com/techprimate/TPPDF/blob/master/Shared/Examples/HeaderFooterExampleFactory.swift):

```
func generateDocument() -> [PDFDocument] {
    let document = PDFDocument(format: .a4)

    // footer will be offset by the height that is equal to the header height
    document.add(.headerRight, textObject: PDFSimpleText(text: "Header right 1"))

    document.add(.footerLeft, textObject: PDFSimpleText(text: "Footer Left 1"))
    document.add(.footerLeft, textObject: PDFSimpleText(text: "Footer Left 2"))
    document.add(.footerLeft, textObject: PDFSimpleText(text: "Footer Left 3"))
    document.add(.footerLeft, textObject: PDFSimpleText(text: "Footer Left 4"))

    var items = [[String?]]()

    for _ in 0..<200 {
        items.append(["Lorem Ipsum is simply dummy text of the printing and typesetting industry."])
    }

    let table = PDFTable(rows: items.count, columns: 1)

    table.content = items
    table.rows.allCellsStyle = PDFTableCellStyle(colors: (fill: .clear, text: .black))
    table.widths = [1]

    table.padding = 10.0

    document.add(table: table)

    return [document]
}
```

Here's how it looks without the fix:

no_fix

And this is with the fix:

Fix

**Comment by philprime** [comment]
Thanks a lot for the example! Helps massively.
I will look into it a little further as I am still confused where my thought is wrong.

**Comment by philprime** [comment]
I took a closer look, and you were right. The calculation itself doesn't make any sense and seems to be a copy paste error. Anyways I added a suggestion so the method includes all paddings, spacings etc.

If you approve my recommendation we can merge it into `develop` (I will change the branch)

File: Source/Internal/Utils/PDFCalculations.swift [hunk 1]
```diff
@@ -163,7 +163,7 @@ internal enum PDFCalculations {
 
         return pageLayout.height
             - layout.margin.top
-            - layout.heights.maxHeaderHeight()
+            - layout.heights.maxFooterHeight()
     }
```
Comment (philprime) [inline Source/Internal/Utils/PDFCalculations.swift:167]:
```suggestion
    /// Calculates the maximum offset from the top edge when the main content should break to the next page
    ///
    /// This methods calculates the limit by the following formula:
    ///
    ///       page height
    ///     - bottom margin
    ///     - footer height
    ///     - footer spacing (if footer exists)
    ///     - padding
    ///     -----------------------------------
    ///       offset from top edge
    ///
    ///     --- ┏━━━━━━━━━━┓
    ///      ↑  ┃┌────────┐┃
    ///      ↓  ┃│ Group  │┃
    ///     --> ┃└────────┘┃
    ///         ┠┄┄┄┄┄┄┄┄┄┄┨
    ///         | spacing  ┃
    ///         ┠┄┄┄┄┄┄┄┄┄┄┨
    ///         ┃  footer  ┃
    ///         ┠┄┄┄┄┄┄┄┄┄┄┨
    ///         ┃  margin  ┃
    ///         ┗━━━━━━━━━━┛
    ///
    /// - Parameter generator: Generator currently in use holding information about the document
    /// - Returns: Offset from top edge in points
    internal static func calculateBottomMaximum(for generator: PDFGenerator) -> CGFloat {
        let layout = generator.layout
        let pageLayout = generator.document.layout
        let footerHeight = layout.heights.maxFooterHeight()

        return pageLayout.height
            - generator.currentPadding.bottom
            - (footerHeight > 0 ? pageLayout.space.footer : 0)
            - footerHeight
            - layout.margin.bottom
    }
```

Comment (philprime) [inline Source/Internal/Utils/PDFCalculations.swift:167]: Replace the whole function with the following:

```swift
    /// Calculates the maximum offset from the top edge when the main content should break to the next page
    ///
    /// This methods calculates the limit by the following formula:
    ///
    ///       page height
    ///     - bottom margin
    ///     - footer height
    ///     - footer spacing (if footer exists)
    ///     - padding
    ///     -----------------------------------
    ///       offset from top edge
    ///
    ///     --- ┏━━━━━━━━━━┓
    ///      ↑  ┃┌────────┐┃
    ///      ↓  ┃│ Group  │┃
    ///     --> ┃└────────┘┃
    ///         ┠┄┄┄┄┄┄┄┄┄┄┨
    ///         ┃ spacing  ┃
    ///         ┠┄┄┄┄┄┄┄┄┄┄┨
    ///         ┃  footer  ┃
    ///         ┠┄┄┄┄┄┄┄┄┄┄┨
    ///         ┃  margin  ┃
    ///         ┗━━━━━━━━━━┛
    ///
    /// - Parameter generator: Generator currently in use holding information about the document
    /// - Returns: Offset from top edge in points
    internal static func calculateBottomMaximum(for generator: PDFGenerator) -> CGFloat {
        let layout = generator.layout
        let pageLayout = generator.document.layout
        let footerHeight = layout.heights.maxFooterHeight()

        return pageLayout.height
            - generator.currentPadding.bottom
            - (footerHeight > 0 ? pageLayout.space.footer : 0)
            - footerHeight
            - layout.margin.bottom
    }
```

**Comment by philprime** [comment]
Can you please apply my suggested change and update the PR? I will merge it afterwards and release an update

**Comment by philprime** [comment]
Relates to #249

**Comment by lpeancovschi** [comment]
Amazing, thanks for reviewing this PR!

**Comment by lpeancovschi** [comment]
Don't know why the code analysis is failing. Codacy doesn't provide much context.

**Comment by lpeancovschi** [comment]
@philprime are you planning to release #205 too?

**Comment by philprime** [comment]
Don't worrry about Codacy, looks like some project linking issue. They changed their GitHub integration a little while ago so it might be broken.

Yes I am planning to release #223 too

**Pull request merged**